File: General-DSA-Questions/Binary Search Tree/1_createBST.cpp
// Write a function to create a Binary Search Tree (BST).

// Approach for inserting a node into a BST:
// 1. If the root is NULL, create a new node with the given data and return it.
// 2. If the data is greater than the root data, insert the data into the right subtree.
// 3. If the data is less than the root data, insert the data into the left subtree.
// 4. Return the root.

// Time Complexity: O(log N) in the average case and O(N) in the worst case.
// Space Complexity: O(log N) in the average case and O(N) in the worst case.

#include <iostream>
using namespace std;

class Node {
  public:
    int data;
    Node* left;
    Node* right;

    Node(int val) {
        data = val;
        left = NULL;
        right = NULL;
    }
};

// CREATE BINARY SEARCH TREE
Node* inserIntoBST(Node* root, int data)  {
    // Create first node when root is null
    if  (root == NULL)  {
        Node* root = new Node(data);
        return root;
    }

    // Not for first Node, Adding data to right
    if  (data > root->data) {
        // Right Sub Tree
        root->right = inserIntoBST(root->right, data);
    }
    else  {   // Adding data to left
        // Left Sub Tree
        root->left = inserIntoBST(root->left, data);
    }

    return root;
}

void createBST(Node* &root){
    cout<< "Enter Data: " << endl;
    int data;
    cin >> data;

    // -1 is the terminating condition for the input, so if enter -1 we are stopping insertion
    while (data != -1)
    {
        root = inserIntoBST(root, data);
        cin >> data;
    }
}

// Level order traversal in level wise
void levelOrderTraversalLevelWise(Node* root){
    queue<Node*> q;
    q.push(root);
    q.push(NULL);

    while(!q.empty()){
        Node* frontNode = q.front();
        q.pop();
        
        if(frontNode == NULL){
            cout<<endl;
            if(!q.empty()){ 
                q.push(NULL);
            }
        }
        else{
            cout<< frontNode->data << " ";

            if(frontNode->left != NULL){
                q.push(frontNode->left);
            }
            if(frontNode->right != NULL){
                q.push(frontNode->right);
            }
        }
    }
}

int main(){
    Node* root = NULL;
    createBST(root);

    cout << "Level Wise Order: " << endl;
    levelOrderTraversalLevelWise(root);

    return 0;
}

/*
Binary Tree Input: 50 30 40 20 60 55 70 80 25 -1

OUTPUT:
Level Wise Order: 
50 
30 60 
20 40 55 70 
25 80 
*/


// Search value in BST
// https://takeuforward.org/data-structure/search-in-a-binary-search-tree-2/ -> Follow this link for searching in a BST

// Approach for searching in a BST:
// 1. If the root is NULL, return NULL.
// 2. If the data is equal to the root data, return the root.
// 3. If the data is greater than the root data, search in the right subtree.
// 4. If the data is less than the root data, search in the left subtree.

// Time Complexity: O(log N) in the average case and O(N) in the worst case.
// Space Complexity: O(log N) in the average case and O(N) in the worst case.

// Function to search for a value in a BST
// Node* searchInBST(Node* root, int data) {
//     // If the root is NULL or the data is found, return the root
//     if (root == NULL || root->data == data) {
//         return root;
//     }

//     // If the data is greater than the root data, search in the right subtree
//     if (data > root->data) {
//         return searchInBST(root->right, data);
//     }

//     // If the data is less than the root data, search in the left subtree
//     return searchInBST(root->left, data);
// }

// But note that recursion requires extra space in the form of a function call stack. We can also solve this problem iteratively using a while loop. Iterative solutions are more efficient in terms of space complexity.

// Iterative approach to search for a value in a BST
// Node* searchInBST(Node* root, int data) {
//     // Traverse the tree until the root is NULL
//     while (root != NULL) {
//         // If the data is found, return the root
//         if (root->data == data) {
//             return root;
//         }

//         // If the data is greater than the root data, search in the right subtree
//         if (data > root->data) {
//             root = root->right;
//         } else {
//             // If the data is less than the root data, search in the left subtree
//             root = root->left;
//         }
//     }

//     // If the data is not found, return NULL
//     return NULL;
// }

File: JS-DSA-Questions/Binary Search Tree/1_createBST.js
// Create a Binary Search Tree (BST) and traverse it level by level.

// Insert approach:
// 1. If root is null, create and return a new node.
// 2. If data > root.data, insert into the right subtree.
// 3. If data < root.data, insert into the left subtree.
// 4. Return the root.
// Time: O(log N) average, O(N) worst. Space: O(log N) average, O(N) worst.

class Node {
  constructor(val) {
    this.data = val;
    this.left = null;
    this.right = null;
  }
}

function insertIntoBST(root, data) {
  if (root === null) {
    return new Node(data);
  }

  if (data > root.data) {
    root.right = insertIntoBST(root.right, data);
  } else {
    root.left = insertIntoBST(root.left, data);
  }

  return root;
}

// C++ read values from cin until -1. Here we build the BST from an array (terminated by -1).
function createBST(values) {
  let root = null;
  for (const data of values) {
    if (data === -1) break;
    root = insertIntoBST(root, data);
  }
  return root;
}

// Level-order traversal, printing each level on its own line.
// JavaScript note: C++ used std::queue and pushed NULL as a level separator; here we use a plain
// array as a queue and the same NULL(-marker) trick with null.
function levelOrderTraversalLevelWise(root) {
  if (root === null) return;
  const q = [root, null];

  let line = [];
  while (q.length > 0) {
    const frontNode = q.shift();
    if (frontNode === null) {
      console.log(line.join(" "));
      line = [];
      if (q.length > 0) q.push(null);
    } else {
      line.push(frontNode.data);
      if (frontNode.left !== null) q.push(frontNode.left);
      if (frontNode.right !== null) q.push(frontNode.right);
    }
  }
}

// Search a value in a BST (recursive)
function searchInBST(root, data) {
  if (root === null || root.data === data) return root;
  return data > root.data ? searchInBST(root.right, data) : searchInBST(root.left, data);
}

// Driver code — input: 50 30 40 20 60 55 70 80 25 -1
const root = createBST([50, 30, 40, 20, 60, 55, 70, 80, 25, -1]);
console.log("Level Wise Order:");
levelOrderTraversalLevelWise(root);
/*
50
30 60
20 40 55 70
25 80
*/
